doc/api/esm.md

```diff
@@ -1129,30 +1129,54 @@ node --experimental-top-level-await b.mjs # works
 
 <!-- type=misc -->
 
+Loaders are modules which export functions and data which can be used by
+Node.js to modify the behavior of loading ES Modules.
+
 To customize the default module resolution, loader hooks can optionally be
 provided via `--experimental-loader ./loader-name.mjs` arguments to Node.js.
 When hooks are used they only apply to ES module loading and not to any
 CommonJS modules loaded.
 
-Loaders are modules which export functions and data which can be used by
-Node.js to modify the behavior of loading ES Modules. Loaders may be ”chained”
-together by passing multiple `--experimental-loader` arguments to Node.js. A
-loader module will be imported once as a singleton.
+A loader module will be imported once as a singleton.
 
-A hook has access to an additional `nextLoader` argument, which contains
-functions which call the hooks of the next loader. When calling the next hook,
-`nextLoader` **does not** need to be passed.
+### Loader Chaining
 
-### <code>nextLoader</code>
+Loaders may be ”chained” together by passing multiple `--experimental-loader`
+arguments to Node.js.
 
-* {Object}
-  * `resolve` {Function}
-  * `getFormat` {Function}
-  * `getSource` {Function}
+There are three types of chaining used in Node.js loaders:
+
+1. First return (`resolve`, `getFormat`, `getSource`): The first hook in the
+  sequence of loaders to return a result is used, falling back to the default
+  Node.js implementation.
+2. Waterfall (`transformSource`): The last return value is used as the input
+  value to the next hook. This allows composition of source transformations.
+3. Direct call (`getGlobalPreloadCode`): These are hooks which are called
+  for all loaders, with all the separate results being individually used.
+
+### Loader Context
+
+In addition to the loader hooks, loaders are provided a `this` context in all
+hook function calls which provides each of the complete hooks as a function:
+
+* `resolve` {Function}
+* `getFormat` {Function}
+* `getSource` {Function}
+* `transformSource` {Function}
+
+And the base Node.js implementations are also provided as the default versions
+of these functions:
+
+* `defaultResolve` {Function}
+* `defaultGetFormat` {Function}
+* `defaultGetSource` {Function}
+
+The argument for each of the above are the same as the standard hook function
+arguments.
```

Comment:
Let's provide `defaultTransform` and `defaultGetGlobalPreloadCode` too, even though they're no-ops. Not that they ever won't be no-ops, but the API feels incomplete without them.

Reply:
default transform is a *maybe*, I'm pretty strongly against that, but a default global preload code is just nonsensical.